helix-term/src/ui/picker.rs

```diff
@@ -123,70 +124,120 @@ impl Preview<'_, '_> {
+pub struct Injector<T, D> {
     dst: nucleo::Injector<T>,
-    editor_data: Arc<T::Data>,
-    shutown: Arc<AtomicBool>,
+    columns: Arc<Vec<Column<T, D>>>,
+    editor_data: Arc<D>,
+    version: usize,
+    picker_version: Arc<AtomicUsize>,
 }
 
-impl<T: Item> Clone for Injector<T> {
+impl<I, D> Clone for Injector<I, D> {
     fn clone(&self) -> Self {
         Injector {
             dst: self.dst.clone(),
+            columns: self.columns.clone(),
             editor_data: self.editor_data.clone(),
-            shutown: self.shutown.clone(),
+            version: self.version,
+            picker_version: self.picker_version.clone(),
         }
     }
 }
 
+#[derive(Debug)]
 pub struct InjectorShutdown;
 
-impl<T: Item> Injector<T> {
+impl<T, D> Injector<T, D> {
     pub fn push(&self, item: T) -> Result<(), InjectorShutdown> {
-        if self.shutown.load(atomic::Ordering::Relaxed) {
+        if self.version != self.picker_version.load(atomic::Ordering::Relaxed) {
             return Err(InjectorShutdown);
         }
 
-        if let Some((item, matcher_text)) = item_to_nucleo(item, &self.editor_data) {
-            self.dst.push(item, |dst| dst[0] = matcher_text);
-        }
+        inject_nucleo_item(&self.dst, &self.columns, item, &self.editor_data);
         Ok(())
     }
 }
 
-pub struct Picker<T: Item> {
-    editor_data: Arc<T::Data>,
-    shutdown: Arc<AtomicBool>,
+impl std::fmt::Display for InjectorShutdown {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "picker has been shut down")
+    }
+}
+
+impl std::error::Error for InjectorShutdown {}
+
+type ColumnFormatFn<T, D> = for<'a> fn(&'a T, &'a D) -> Cell<'a>;
+
+pub struct Column<T, D> {
+    name: &'static str,
+    format: ColumnFormatFn<T, D>,
+    /// Whether the column should be passed to nucleo for matching and filtering.
+    /// `DynamicPicker` uses this so that the dynamic column (for example regex in
+    /// global search) is not used for filtering twice.
+    filter: bool,
+}
+
+impl<T, D> Column<T, D> {
+    pub fn new(name: &'static str, format: ColumnFormatFn<T, D>) -> Self {
+        Self {
+            name,
+            format,
+            filter: true,
+        }
+    }
+
+    pub fn without_filtering(mut self) -> Self {
+        self.filter = false;
+        self
+    }
+
+    fn format<'a>(&self, item: &'a T, data: &'a D) -> Cell<'a> {
+        (self.format)(item, data)
+    }
+
+    fn format_text<'a>(&self, item: &'a T, data: &'a D) -> Cow<'a, str> {
+        let text: String = self.format(item, data).content.into();
+        text.into()
+    }
+}
+
+/// Returns a new list of options to replace the contents of the picker
+/// when called with the current picker query,
+type DynQueryCallback<T, D> =
+    fn(&str, &mut Editor, Arc<D>, &Injector<T, D>) -> BoxFuture<'static, anyhow::Result<()>>;
+
+pub struct Picker<T: 'static + Send + Sync, D: 'static> {
+    column_names: Vec<&'static str>,
```

Comment:
We discussed this privately. We allocate for the column but also per query (the PickerQuery type owns these strings too). Since they're immutable and we're cloning them every now and then `Arc<str>` is a good choice. Plugins would still be able to use custom column names because they aren't `'static`